proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,754 @@
+        // Key is present (matches any value except for the key being absent).
+        Exists exists = 3;
+      }
+      // If set to true, the match is inverted -- i.e., the key must NOT
+      // match the specified value.
+      bool invert = 4;
+    }
+
+    enum MatchType {
+      // Default value.
+      MATCH_TYPE_UNSPECIFIED = 0;
+      // Logical AND of constraints.
+      MATCH_TYPE_AND = 1;
+      // Logical OR of constraints.
+      MATCH_TYPE_OR = 2;
+    }
+
+    // A list of key/value constraints.
+    repeated Constraint constraints = 1;
+
+    // How to match the constraints.
+    MatchType match_type = 2;
+  }
+
+  // A list of constraint lists. All constraint lists must match (i.e.,
+  // logical AND semantics).
+  repeated ConstraintList constraints = 1;
+}
+```
+
+#### Where Matching Is Performed
+
+Both xDS servers and clients need to evaluate matching between a set
+of dynamic parameters and a set of constraints.  The server does this
+when deciding which variant of a given resource to return for a given
+subscription request.  When the client receives the resource from
+the server, it needs to do the same matching to determine which of its
+subscriptions that resource is associated with.  Therefore, the matching
+behavior becomes an inherent part of the xDS transport protocol.
+
+Note that because leaf clients should only ever receive a single variant
+of a given resource, implementations may be tempted to not bother with
+this matching on the client side.  However, that is not true for caching
+xDS proxies; a proxy may have multiple clients that request different
+variants of the same resource.  Because we do not want to get into a
+situation where xDS servers typically do not populate dynamic parameter
+constraints in their responses and then need changes to work with
+caching proxies, this design requires that leaf clients validate that the
+constraints in the response match the requested dynamic parameters.
+This ensures that the wire protocol used by leaf clients and caching xDS
+proxies remains the same.
+
+(In effect, the resource cache in an xDS client is basically the same
+logic as that on an xDS server; the only difference is that in the case
+of a client, the resources in the cache come from an xDS stream instead
+of from an authoritative database.  Similarly, a caching xDS proxy is
+simply an xDS client where the subscriptions come from an incoming xDS
+stream.)
+
+#### Example: Basic Dynamic Parameters Usage
+
+Let's say that the clients are currently categorized by the parameter
+`env`, whose value is either `prod` or `test`.  So any given client will
+send one of the following sets of dynamic parameters:
+- `{env=prod}`
+- `{env=test}`
+
+Now let's say that the server has two variants of a given resource, and
+the variants have the following dynamic parameter constraints:
+
+```textproto
+// For {env=prod}
+{constraints:[
+  {
+    constraints:[{key:"env" value:"prod"}]
+    match_type: MATCH_TYPE_AND
```

Comment:
nit: it's kind of confusing that we specify AND here with a single constraint. Maybe remove and/or say this doesn't need setting with a single value? Up to you.

Reply:
I think this is related to the question above about whether `MATCH_TYPE_AND` is actually useful at all.  If it's not, we can remove it, but I'd like @fuqianggao to weigh in on this first, since it was based on an internal proto that he showed me.